### PR TITLE
Upgrade to Rails 5.1 and Ruby 2.4

### PR DESCRIPTION
**Pull request by etdsoft**
The main interesting parts were:

* We had to bump `json` gem ahead of Rails 5.1 requirements for Rails 2.4 compatibility.
* We had to bump `paper_trail` to the 6.x branch to make it compatible with AR 5.1
* An ActionDispatch exception class name had to be updated.

The `favicon.ico` change is unrelated to the upgrade, but should deal with the 404s in production.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -151,7 +151,7 @@ group :development do
   # listen higher than 3.1.1 require Ruby version >= 2.2.3 (we're currently on
   # 2.2.2). Restrict the version of `listen` to prevent `guard-rspec`
   # introducing an incompatible dependency:
-  gem 'listen', '~> 3.0.5', '<= 3.1.1'
+  gem 'listen', '>= 3.0.5', '< 3.2'
```
Comment (dormi) [inline Gemfile]: I have this error when running `bundle`:
```
Gem::InstallError: ruby_dep requires Ruby version >= 2.2.5, ~> 2.2.
An error occurred while installing ruby_dep (1.5.0), and Bundler cannot continue.
Make sure that `gem install ruby_dep -v '1.5.0'` succeeds before bundling.
```
It works when changing this line with:
`gem 'listen', '~> 3.0.8'`

Comment (etdsoft) [inline Gemfile]: @dormi did you spot this change to `.ruby-version` ?

https://github.com/dradis/dradis-ce/pull/149/files#diff-0d75b3c58057105d6c02c1fd3827050d

Comment (dormi) [inline Gemfile]: Of course, missed that, sorry.

Comment (dormi) [inline Gemfile]: Once in ruby `2.4.1`, I'm see this warning when booting rails:
```
[...]/.rvm/gems/ruby-2.4.1/gems/sinatra-2.0.0.beta2/lib/sinatra/base.rb:1256: warning: constant ::Fixnum is deprecated
```

Do you see it?

Comment (dormi) [inline Gemfile]: can we change  `gem 'sinatra', '2.0.0.beta2'` with `gem 'sinatra'`? or just remove the line?

Comment (etdsoft) [inline Gemfile]: yes, most likely due to resque requirements:

```
# Redis-based background worker
gem 'resque', require: 'resque/status_server'
gem 'resque-status'
# See https://github.com/sinatra/sinatra/issues/1055
gem 'sinatra', '2.0.0.beta2'
```

Independent of this PR.

Comment (etdsoft) [inline Gemfile]: if `gem 'sinatra'` deals with it, feel free to push the change.

**Pull request merged**